osbuild/objectstore.py

```diff
@@ -100,25 +100,41 @@ def new(self, object_id, base_id=None):
 
             yield tree.path
 
-            # if `tree.treesum` raises an exception, the working tree
-            # is cleaned up by tempfile, otherwise, we save it in the
-            # correct place:
-            treesum_hash = tree.treesum
-
-            # the tree is stored in the objects directory using its content
-            # hash as its name, ideally a given object_id (i.e., given config)
-            # will always produce the same content hash, but that is not
-            # guaranteed
-            output_tree = f"{self.objects}/{treesum_hash}"
-
-            # if a tree with the same treesum already exist, use that
-            with suppress_oserror(errno.ENOTEMPTY):
-                os.rename(tree, output_tree)
-
-            # symlink the object_id (config hash) in the refs directory to the
-            # treesum (content hash) in the objects directory. If a symlink by
-            # that name alreday exists, atomically replace it, but leave the
-            # backing object in place (it may be in use).
+            # Now the tree was created or modified by the caller all
+            # that is left to do is to write it to the object store
+            self.commit(tree, object_id)
+
+    def commit(self, tree: TreeObject, object_id: str) -> str:
+        """Commits a TreeObject to the object store
+
+        Move the contents of the tree (TreeObject) to object directory
+        of the store with the content hash (tree.treesum) as its name.
+        Creates a symlink to that ('objects/{hash}') in the references
+        directory with the object_id as the name ('refs/{object_id}).
+        If the link already exists, it will be atomically replaced.
+
+        Returns: The treesum of the tree
+        """
+        treesum_hash = tree.treesum
+
+        # the tree is stored in the objects directory using its content
+        # hash as its name, ideally a given object_id (i.e., given config)
+        # will always produce the same content hash, but that is not
+        # guaranteed
+        output_tree = f"{self.objects}/{treesum_hash}"
+
+        # if a tree with the same treesum already exist, use that
+        with suppress_oserror(errno.ENOTEMPTY):
+            os.rename(tree.path, output_tree)
+            tree.path = output_tree
```

Comment:
Hm, what is the intended semantics of `tree.path`? I would have expected it to always be set to `output_tree`, even if the rename fails (so the assignment to happen outside the `with` clause). If the rename fails it means that we should still use the existing `output_tree` and discard the one we created. If I understand correctly, that would just work with this code, as `tree.path` is in a tmpdir, so if we just forget about it it gets cleaned up for us (maybe do the cleanup explicitly in this helper to make it clearer/more robust?).

Reply:
Ninja edit: you fix this in a following patch. Suggestion to delete the original still stands.